encoding/ccf/encode.go

```diff
@@ -0,0 +1,1751 @@
+/*
+ * Cadence - The resource-oriented smart contract programming language
+ *
+ * Copyright 2022-2023 Dapper Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ccf
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	goRuntime "runtime"
+	"sort"
+	"sync"
+
+	"github.com/fxamacker/cbor/v2"
+	"github.com/onflow/cadence"
+	"github.com/onflow/cadence/runtime/common"
+)
+
+// CBOREncMode
+//
+// See https://github.com/fxamacker/cbor:
+// "For best performance, reuse EncMode and DecMode after creating them."
+var CBOREncMode = func() cbor.EncMode {
+	options := cbor.CoreDetEncOptions()
+	options.BigIntConvert = cbor.BigIntConvertNone
+	encMode, err := options.EncMode()
+	if err != nil {
+		panic(err)
+	}
+	return encMode
+}()
+
+// An Encoder converts Cadence values into CCF-encoded bytes.
+type Encoder struct {
+	// CCF codec uses CBOR codec under the hood.
+	enc *cbor.StreamEncoder
+}
+
+// Encode returns the CCF-encoded representation of the given value.
+//
+// This function returns an error if the Cadence value cannot be represented in CCF.
+func Encode(value cadence.Value) ([]byte, error) {
```

Comment:
CCF encoder uses the same API as the encoder for JSON-Cadence Data Interchange:
- `Encode(cadence.Value)`
- `MustEncode(cadence.Value)`
- `NewEncoder(io.Writer)`
- `Encoder.Encode(cadence.Value)`